[APM] Replace environment dropdown with SuggestionsSelect in landing pages and service overview page

kpatticha opened the pull request:
Part of https://github.com/elastic/kibana/issues/117535

This PR replaces the environment dropdown selection with the `SuggestionsSelect` which uses `term_enum` in landing page and `EnvironmentSelect` in service overview.  I created a new endpoint for the latter case cause I had to do an aggregation in order to get the `service.environment`
 
### UI 
- Replace `get_environments`  with `suggestions`
- Replace `get_service_environments` with `service_environment_suggestions_select` 
- Introduce new component `ServiceEnvironmentSuggestionsSelect` 

### Backend
- Introduce new api `service_environment_suggestions_select` 
- Introduce new api `has_not_defined_environment` 
- Extend `internal/apm/suggestions` to accept range query params


### Example


https://user-images.githubusercontent.com/3369346/159377304-8095207a-e2ad-4449-b71d-2881fadc5c03.mov



### Notes

#### it shows "Not defined" option only if there is a service with a not defined environment. 

### Todo
- [x] bring back not defined 
- [x] Remove unused endpoint/ code 
- [x] DRY - duplicate code , find a better way to share code between `ServiceEnvironmentSuggestionsSelect` and `SuggestionsSelect` UI component

kpatticha committed:
Extend /suggestions api to accept range query

kpatticha committed:
Add new endpoint /suggestions_with_terms

kpatticha committed:
Use suggestionSelect for environment filter

sqren commented on x-pack/plugins/apm/server/routes/suggestions/get_suggestions.ts:
```diff
@@ -37,6 +42,10 @@ export async function getSuggestions({
       string,
+      ...(start &&
+        end && {
+          index_filter: rangeQuery(start, end)[0],
```
This doesn't read super well. Not sure if we should just not use the `rangeQuery` helper

kpatticha replied:
https://github.com/elastic/kibana/pull/126679/commits/760f49a445560b6c8052a257a1a555ea67a86a88

sqren commented on x-pack/plugins/apm/common/anomaly_detection/get_preferred_service_anomaly_timeseries.ts:
```diff
@@ -12,13 +12,11 @@ import { ServiceAnomalyTimeseries } from './service_anomaly_timeseries';
 
 export function getPreferredServiceAnomalyTimeseries({
```
This function is only used by `usePreferredServiceAnomalyTimeseries`. It'll be easier to read if we just inline it.

sqren commented on x-pack/plugins/apm/common/anomaly_detection/get_preferred_service_anomaly_timeseries.ts:
```diff
@@ -28,9 +26,7 @@ export function getPreferredServiceAnomalyTimeseries({
-    environment === ENVIRONMENT_ALL.value && environments.length === 1
-      ? environments[0]
-      : environment;
+    environment === ENVIRONMENT_ALL.value ? ENVIRONMENT_ALL.label : environment;
```
We had some logic where we wanted to select a specific environment if there was only one. Do we still do that?

sqren replied:
Btw. if you want the environment label you should use `getEnvironmentLabel(environment)`

kpatticha replied:
hm no cause previously the `EnvironmentsContextProvider` was responsible for fetching the environments. Now `SuggestionsSelect` component is fetching the data so the `environments` are not available to the `EnvironmentsContextProvider`. 

if we have to keep and want to keep the same logic I need to figure out how to share the results from the SuggestionsSelect 🤔

kpatticha replied:
I moved that logic to the suggestion select component. if there is only one option available it will select this one.

https://github.com/elastic/kibana/pull/126679/commits/34750b7e8fa52dc296cb349b90d4bf6f77efb06b

formgeist commented on x-pack/plugins/apm/public/components/shared/suggestions_select/suggestions_select_with_term.tsx:
```diff
@@ -0,0 +1,125 @@
+  return (
+    <EuiComboBox
+      async={true}
+      compressed={true}
```
@kpatticha Any way we can pass this as `false` in the case of the environment selector on the service views? We want the control to be the default size, not compact.

I imagine this is the same component in use in the anomaly detection job creation?
<img width="538" alt="CleanShot 2022-03-02 at 20 17 40@2x" src="https://user-images.githubusercontent.com/4104278/156433017-8648bd41-cf7b-4361-9e35-9df5b103b3fb.png">

kpatticha replied:
For `EuiComboBox` the `compressed` default value is `false` so I'm removing it. We can customise later or if needed. 

Both are using the `EuiComboBox` component.

formgeist reviewed (commented):
Left a comment on the size of the `EuiComboBox`

Another thing I found was that the component comes with a fixed `clearable` option so it will always allow the user to clear the input. As opposed to the environment selection when creating e.g. ML jobs, we actually want the component to function more as a regular `select` form field where the user can only choose the options. This obviously then conflict with the type-ahead functionality that the `EuiComboBox` provides, so perhaps it's a question on whether or not the component itself should offer an option to disable the `clearable` action.

The issue that the user might run into is that they can clear the input and then what's being shown?

![CleanShot 2022-03-02 at 20 24 57](https://user-images.githubusercontent.com/4104278/156434337-103b7877-2137-4e97-a729-cf864b81cb2c.gif)

sqren commented on x-pack/plugins/apm/public/components/shared/environment_filter/index.tsx:
```diff
@@ -34,76 +38,88 @@ function updateEnvironmentUrl(
-      text: environment,
-    }));
-
-  return [
-    ENVIRONMENT_ALL,
-    ...(environments.includes(ENVIRONMENT_NOT_DEFINED.value)
-      ? [ENVIRONMENT_NOT_DEFINED]
-      : []),
-    ...(environmentOptions.length > 0 ? [SEPARATOR_OPTION] : []),
-    ...environmentOptions,
-  ];
```
Can the user still select "Not defined" and "All" environments?

kpatticha replied:
@sqren considering this https://github.com/elastic/kibana/issues/71239 do you have any concern about removing the "Not defined" option from the dropdown?

kpatticha replied:
we have 2 options:

1. Hard code the "Not defined" option and display it always 
2. Create a new search to determine if there is a `service` without a `service.environment`  and based on the results add the "Not defined" option. 
For example, 
 ```
{
  "size": 0,
  "terminate_after": 1,
  "query": {
    "bool": {
      "must_not": [
        {"exists": {"field": "service.environment"}}
      ]
    }
  }
}
```
With option 2 we mimic the experience we provide today, the only downside is one more request but it should be fast

sqren replied:
I think it's better if you can do option 2. It should be fast enough.
But won't it cause problems for your suggestion component that now have to be concerned with a secondary request?

sqren commented on x-pack/plugins/apm/public/components/shared/environment_filter/index.tsx:
```diff
@@ -34,76 +38,88 @@ function updateEnvironmentUrl(
+      }
+      defaultValue={getEnvironmentLabel(environment)}
+      field={SERVICE_ENVIRONMENT}
+      termField={SERVICE_NAME}
+      termValue={termValue}
```
Should this perhaps be:

```suggestion
      fieldName={SERVICE_NAME}
      fieldValue={termValue}
```

kpatticha replied:
https://github.com/elastic/kibana/pull/126679/commits/4eae34b4ad02be647d0e51c565a706104c85200c

sqren commented on x-pack/plugins/apm/server/routes/suggestions/route.ts:
```diff
@@ -30,16 +38,61 @@ const suggestionsRoute = createApmServerRoute({
+});
+
+const suggestionsWithTermsRoute = createApmServerRoute({
+  endpoint: 'GET /internal/apm/suggestions_with_terms',
+  params: t.type({
+    query: t.intersection([
+      t.type({
+        field: t.string,
+        string: t.string,
+        termField: t.string,
+        termValue: t.string,
+      }),
+      rangeRt,
```
I'm not sure about this very generic endpoint. But if we are going to do it I think it should take a `filter` instead:

```suggestion
  endpoint: 'GET /internal/apm/suggestions_with_filter',
  params: t.type({
    query: t.intersection([
      t.type({
        field: t.string,
        string: t.string,
        filter: t.string
      })
```

sqren replied:
And to stay consistent with my other comments, I also think we should rename `field` -> `fieldName` and `string` -> `fieldValue`

```suggestion
  endpoint: 'GET /internal/apm/suggestions_with_filter',
  params: t.type({
    query: t.intersection([
      t.type({
        fieldName: t.string,
        fieldValue: t.string,
        filter: t.string
      })
```

kpatticha replied:
@sqren 

> I'm not sure about this very generic endpoint. But if we are going to do it I think it should take a `filter` instead:

I had my doubts about the generic endpoint. IMO it was too implicit with no obvious benefits and currently the only place where I'm using it is in the service environment filter. 

so instead I created the [getSuggestionsByServiceName](https://github.com/elastic/kibana/commit/4eae34b4ad02be647d0e51c565a706104c85200c#diff-741fb8e4f5601e16766676f59428e73622c49f33ab1b58e8bfbc53c823225617R12). it should be more clear what it does

kpatticha replied:
https://github.com/elastic/kibana/pull/126679/commits/4eae34b4ad02be647d0e51c565a706104c85200c

sqren replied:
If we only use it for environments, and since you've already hardcoded `service.name` as the only filter, I'm inclined to say that it should be renamed from `getSuggestionsByServiceName` to `getEnvironmentSuggestions`

sqren replied:
(and then `fieldName` should be replaced with `service.environment` hardcoded)

sqren commented on x-pack/plugins/apm/server/routes/suggestions/get_suggestions_with_term.ts:
```diff
@@ -0,0 +1,66 @@
+  const response = await apmEventClient.search('get_suggestions_with_terms', {
+    apm: {
+      events: [
+        ProcessorEvent.transaction,
+        ProcessorEvent.error,
+        ProcessorEvent.metric,
```
Does it make sense for a generic endpoint to have these indices hardcoded?

kpatticha replied:
I'm quite sure, I used the same as `/suggestions` endpoint.

sqren replied:
Okay, let's leave it for now. We might have to change this as we see more consumers.

sqren commented on x-pack/plugins/apm/server/routes/suggestions/get_suggestions_with_term.ts:
```diff
@@ -0,0 +1,66 @@
+      size: 0,
+      query: {
+        bool: {
+          filter: [
+            ...filter,
+            ...(start && end ? rangeQuery(start, end) : []),
+            {
+              query_string: {
+                fields: [field],
+                query: `*${string}*`,
+              },
+            },
+          ],
```
Two things: 
 - what about using the `prefix` query? I think that's the behaviour of the terms_enum api that we are mimicking. 
 - Can we rename `field` to `fieldName` and `string` to `fieldValue`. I think we should consider doing the same for our suggestion APIs (I know that's what the terms enum API calls it but we can call it differently in our APIs).

```suggestion
          filter: [
            ...filter,
            prefix: { [fieldName]: fieldValue }
          ],
```

kpatticha replied:
I didn't know the existence of `prefix` query 🙈  but after experimenting a bit with it I believe it won't be as useful as the `query_string `.  

`prefix` returns the documents where the `fieldName` contains a term that **begins** with the `fieldValue`. I doubt the users will search with the termValue `https://` to get their environment which unfortunately are set with the host name.

kpatticha replied:
I tried `wildcard` query and it's much faster

kpatticha replied:
https://github.com/elastic/kibana/pull/126679/commits/ba53e9ea143169addfb522be69d551224522e50a

kpatticha committed:
Clean up route

kpatticha committed:
UI tweaks for the EuiComboBox

kpatticha committed:
Make suggestion endpoint for service enrvironment explicit

- follow same naming convention for fields and values

kpatticha committed:
Fix deps

sqren commented on x-pack/plugins/apm/server/routes/suggestions/get_suggestions_by_service_name.ts:
```diff
@@ -0,0 +1,72 @@
+import { rangeQuery, termQuery } from '../../../../observability/server';
+import { Setup } from '../../lib/helpers/setup_request';
+
+export async function getSuggestionsByServiceName({
```
Good idea changing this from a generic endpoint. Will probably make things easier.

kpatticha committed:
Replace query_string with wildcard query

kpatticha committed:
Rename field to fieldName

kpatticha committed:
Introduce new enpoint has not defined environment

kpatticha committed:
Show Not defined option in environment filter

kpatticha commented on x-pack/plugins/apm/server/routes/suggestions/get_suggestions_by_service_name.ts:
```diff
@@ -0,0 +1,72 @@
+                query_string: {
+                  fields: [fieldName],
+                  query: `*${fieldValue}*`,
+                },
```
https://github.com/elastic/kibana/pull/126679/commits/28dfa240fa3c656ab4c0c1ef96e0d04195e94f14

kpatticha committed:
get environment label for anomaly timeseries

kpatticha committed:
Fix types

kpatticha committed:
Select environment if there is only one available

kpatticha committed:
Fix tests

dgieselaar commented:
> if there is only one environment available it selects this one. This change has its own pros and cons

Was this discussed offline? Curious about the rationale. This has a pretty big impact on UX.

kpatticha commented:
> > if there is only one environment available it selects this one. This change has its own pros and cons
> 
> Was this discussed offline? Curious about the rationale. This has a pretty big impact on UX.

Now the `SuggestionsSelect` is responsible for fetching the list of environments/suggestions, thus the `environments` list is not available from the context anymore and the idea was to move the logic from [get_preferred_service_anomaly_timeseries](https://github.com/elastic/kibana/pull/126679/files#diff-6353899a06dddff7fbaee5d9903eae4ed47e1dc540c0f1b448105c57aabdad15L30) to the environment filter. 

do you think will be a bad experience for the end user? Do you have another suggestion?

dgieselaar commented:
@kpatticha it depends on whether this is a redirect or not. If it's not a redirect I would be worried about us accidentally linking to the implicit environment even though the user has intentionally selected "All". If it is explicit I think it might be a little annoying e.g. when moving from the service inventory to a service and back, when the automatically selected environment carries over into the service inventory. 

I think what I suggested was to still have environments available in the context and having the SuggestionSelect re-use those environments as default values.

sqren commented on x-pack/plugins/apm/public/components/shared/environment_filter/index.tsx:
```diff
@@ -5,106 +5,60 @@
+    <ServiceEnvironmentSelect
+      prepend={prepend}
+      onChange={(changeValue) =>
+        updateEnvironmentUrl(history, location, changeValue)
```
This updates the url when the input is submitted (on enter). But we should only update the url when a specific option is selected.

kpatticha committed:
move getPreferredServiceAnomalyTimeseries

kpatticha commented:
we discussed earlier today that we don't won't to go with the redirect solution.  

Instead, the environment list will be available in the  `environmentContext` and the environment filter will consume the list from the context and the component itself will be responsible for sending a request with the search value

kpatticha committed:
Remove redirect and get environment list from context

kpatticha commented:
@elasticmachine merge upstream

kpatticha commented:
@elasticmachine merge upstream

dgieselaar commented on x-pack/plugins/apm/server/routes/suggestions/route.ts:
```diff
@@ -30,16 +38,58 @@ const suggestionsRoute = createApmServerRoute({
+  params: t.type({
+    query: t.intersection([
+      t.type({
+        fieldName: t.string,
+        fieldValue: t.string,
```
I'm not particularly convinced by fieldName/fieldValue. IMHO `kuery` would be more useful here. `fieldName` and `fieldValue` implies that we can only filter on one field, `kuery` is more flexible and familiar to us, and considering we don't validate fieldName/value other than being a string, I feel that is the better option.

dgieselaar commented on x-pack/plugins/apm/server/routes/suggestions/route.ts:
```diff
@@ -8,20 +8,28 @@
+  params: t.type({
+    query: t.intersection([
+      t.type({
+        fieldName: t.string,
+        fieldValue: t.string,
```
Same here, I'm inclined to say `kuery` is the better option.

dgieselaar replied:
Ah this makes it hard to combine it with the terms enum API, never mind this comment :)

dgieselaar commented on x-pack/plugins/apm/server/routes/suggestions/get_suggestions_by_service_name.ts:
```diff
@@ -0,0 +1,71 @@
+          bool: {
+            filter: [
+              ...termQuery(SERVICE_NAME, serviceName),
+              ...(start && end ? rangeQuery(start, end) : []),
```
I think it's best that we always have a range set - otherwise this will look at _all_ the data which could take a while. If we don't have a time picker I'd suggest something like `now-24h`.

kpatticha replied:
good point here 👍 

I set the default ranges to `now-24h` 

https://github.com/elastic/kibana/pull/126679/commits/bb56b5e256286d4dbd62f4cf17ab40352a26252d

dgieselaar commented on x-pack/plugins/apm/server/routes/observability_overview/has_not_defined_environment.ts:
```diff
@@ -0,0 +1,69 @@
+/*
```
As mentioned elsewhere, I don't think this should be an endpoint, just a utility function to be used in different endpoints.

kpatticha replied:
removed here https://github.com/elastic/kibana/pull/126679/commits/4e2011adefc0c750626206521d48156c842ab76c

dgieselaar commented on x-pack/plugins/apm/public/components/shared/environment_filter/index.tsx:
```diff
@@ -5,106 +5,37 @@
   history: History,
   location: ReturnType<typeof useLocation>,
-  environment: string
+  environment?: string
```
Why is this no longer required?

dgieselaar commented on x-pack/plugins/apm/public/components/shared/environment_select/index.tsx:
```diff
@@ -0,0 +1,86 @@
+    status,
+  } = useEnvironmentsContext();
+
+  const customOptions = useEnvironmentCustomOptions({
+    serviceName,
+    start,
+    end,
+  });
```
it's a little confusing to me what "custom options" mean in this context - can we give it a more meaningful name?

dgieselaar commented on x-pack/plugins/apm/public/components/shared/environment_select/index.tsx:
```diff
@@ -0,0 +1,86 @@
+    end,
+  });
+
+  if (environment) {
```
this should always be set (to a non-empty string)

dgieselaar commented on x-pack/plugins/apm/public/components/shared/environment_select/index.tsx:
```diff
@@ -0,0 +1,86 @@
+export function EnvironmentSelect({
+  onChange,
+}: {
+  onChange: (value?: string) => void;
```
I think it's confusing to have this component emit an onChange event, but not have the environment value as a prop. Ideally it is either a controlled component, taking both an `environment` or `value` prop, and an `onChange` prop, or uncontrolled, having neither.

dgieselaar commented on x-pack/plugins/apm/public/components/shared/environment_filter/index.tsx:
```diff
@@ -5,106 +5,37 @@
-  );
-}
-
-export function EnvironmentFilter({
```
is this not being used by the UX app?

kpatticha replied:
it seems they're using their own implementation 
https://github.com/elastic/kibana/blob/main/x-pack/plugins/ux/public/components/app/rum_dashboard/environment_filter/index.tsx

dgieselaar replied:
Ok, maybe they copied+pasted it over there. Great that we can clean this up now 👍

dgieselaar commented on x-pack/plugins/apm/public/hooks/use_environments_fetcher.tsx:
```diff
@@ -7,54 +7,64 @@
-      if (start && end) {
-        return callApmApi('GET /internal/apm/environments', {
+      if (serviceName) {
+        return callApmApi('GET /internal/apm/suggestions_by_service_name', {
```
It might be better to have this hook exclusively fetch the list of environments, regardless of filter.

kpatticha replied:
I reverted the this change. 

The hooks fetch the list of the environments. (as it was before) the suggestion endpoint is fired only when the user types a string value . 

https://github.com/elastic/kibana/pull/126679/commits/66337b0ba10e64a74121fad24d960a87e4ea7d2c
https://github.com/elastic/kibana/pull/126679/commits/e0aff7c8ebe8571fe911fbc4f797f4a41d3c6d7a

dgieselaar reviewed (commented):
I think my biggest confusion here stems from how we're using different endpoints. What perhaps would be a more clear design is:

- have EnvironmentsSelect use the environments from context when no filter is set
- if a filter is set, call the suggestions API (`GET /internal/apm/suggestions` for the `service.name` field)
- `GET /internal/apm/suggestions` will fallback to using a terms aggregation if a filter is set

formgeist reviewed (approved):
LGTM 👍 I noticed a few issues, possibly to do with the `EuiComboBox` component.

When I select an option in the dropdown, the typing indicator is active. I think the field just should escape itself out of focus, at least it looks that way but the indicator is still there.

![CleanShot 2022-03-23 at 15 55 04](https://user-images.githubusercontent.com/4104278/159739236-52668a97-80f9-400a-aa45-e88a6a1a47e7.gif)

I raised another [issue](https://github.com/elastic/kibana/issues/128257) that the condition options in our rule creation steps are not filtering when the user selects a specific service.

kpatticha committed:
Remove endpoint for not defined option

kpatticha committed:
Suggestions route fallbacks to use terms aggregation when service name is set

kpatticha committed:
Set default ranges (now-24h)

kpatticha committed:
Get environment list from context

kpatticha commented on x-pack/plugins/apm/server/routes/observability_overview/has_not_defined_environment.ts:
```diff
@@ -0,0 +1,69 @@
+          ],
+        },
+      },
+      aggs: {
```
I removed this endpoint anyways but I used the agg at first place to avoid the issue I describe above with the not defined options

https://github.com/elastic/kibana/pull/126679/commits/4e2011adefc0c750626206521d48156c842ab76c

kpatticha commented on x-pack/test/functional/apps/apm/correlations/latency_correlations.ts:
```diff
@@ -66,11 +66,12 @@ export default function ({ getPageObjects, getService }: FtrProviderContext) {
         const fromTime = 'Jul 29, 2019 @ 00:00:00.000';
         const toTime = 'Jul 30, 2019 @ 00:00:00.000';
         await PageObjects.timePicker.setAbsoluteRange(fromTime, toTime);
+        await testSubjects.click('comboBoxInput');
```
I forgot to revert it. it was used when I did the redirect on the service overview page

dgieselaar commented on x-pack/plugins/apm/server/routes/suggestions/route.ts:
```diff
@@ -30,16 +39,31 @@ const suggestionsRoute = createApmServerRoute({
+      : await getSuggestions({
+          fieldName,
+          fieldValue,
+          searchAggregatedTransactions,
```
What are you unsure about specifically? if it is needed?

kpatticha replied:
that's right. I'm not sure If needed here (or when it's needed)

dgieselaar replied:
I think it's okay to use it here. It will skip the transaction indices if metrics are being used. There's not much of a performance gain but I think it's more _correct_ to ignore transaction data, although the same environments should be available on transaction metrics.

dgieselaar commented on x-pack/plugins/apm/server/routes/suggestions/get_suggestions_with_terms_aggregation.ts:
```diff
@@ -0,0 +1,72 @@
+              ...rangeQuery(start, end),
+              {
+                wildcard: {
+                  [fieldName]: `*${fieldValue}*`,
```
should we use the leading asterisk here? [the docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-wildcard-query.html) mention the following:

> Avoid beginning patterns with * or ?. This can increase the iterations needed to find matching terms and slow search performance.

kpatticha replied:
IMO we should use it cause _unfortunately_ we have cases where they pass a host name as environment value and without the leading asterisk the search experience will be poor. 

For example, I have this environment `https://www.elastic.co` and if I try to search for `elas` without the leading asterisk the results will be empty. I doubt the user will search for `https` to get `https://www.elastic.co`

dgieselaar replied:
I've thought about this a bit, and I feel we should add a `timeout` here, something like 1500ms. These results don't need to be accurate and it's better to return possibly incomplete suggestions in a relatively short time frame rather than having the user potentially wait double-digit seconds before getting suggestions. There's more information about the `timeout` property here. https://www.elastic.co/guide/en/elasticsearch/reference/current/search-search.html

dgieselaar replied:
@kpatticha did you see this comment? I don't see the timeout being added.

kpatticha replied:
sorry for missing your comment @dgieselaar, it's resolved here https://github.com/elastic/kibana/pull/126679/commits/33ab81bd697d911c5807fbc5ba3cc0d26dd04874

dgieselaar commented on x-pack/plugins/apm/server/routes/suggestions/get_suggestions_with_terms_aggregation.ts:
```diff
@@ -0,0 +1,72 @@
+    {
+      apm: {
+        events: [
+          ProcessorEvent.transaction,
```
this is probably where we should use `searchAggregatedTransactions`. If we can avoid iterating over all the transaction events that'd be nice.

kpatticha replied:
updated: https://github.com/elastic/kibana/pull/126679/commits/85d56cbc9818d6bc4c8f962aae641bb5b35d4ed1

dgieselaar commented on x-pack/plugins/apm/server/routes/suggestions/get_suggestions_with_terms_aggregation.ts:
```diff
@@ -0,0 +1,72 @@
+  serviceName,
+  setup,
+  size,
+  start = moment().subtract(24, 'h').valueOf(),
```
I'd prefer for this to be made more explicit - making the route query params required and having the consumer specify them. I think setting default values in the route makes it easy to forget about it.

kpatticha replied:
updated https://github.com/elastic/kibana/pull/126679/commits/30465933cd2bddbf6a26abe0eb9386fbd44985df

dgieselaar commented on x-pack/plugins/apm/server/routes/suggestions/get_suggestions.ts:
```diff
@@ -4,23 +4,27 @@
   setup: Setup;
   size: number;
-  string: string;
+  start?: number;
```
same here, let's make it required on the server if possible.

dgieselaar commented on x-pack/plugins/apm/public/hooks/use_preferred_service_anomaly_timeseries.ts:
```diff
@@ -5,12 +5,43 @@
 import { useEnvironmentsContext } from '../context/environments_context/use_environments_context';
 import { useServiceAnomalyTimeseriesContext } from '../context/service_anomaly_timeseries/use_service_anomaly_timeseries_context';
 
+export function getPreferredServiceAnomalyTimeseries({
```
if this is exported, can it be a separate file?

kpatticha replied:
separeted: https://github.com/elastic/kibana/pull/126679/commits/2d8f66c0e908796e4e4cbfd835ccd3eb1df84eed

dgieselaar commented on x-pack/plugins/apm/public/hooks/use_environments_fetcher.tsx:
```diff
@@ -34,7 +40,7 @@ export function useEnvironmentsFetcher({
   end?: string;
 }) {
-  const { data = INITIAL_DATA, status } = useFetcher(
+  const { data = INITIAL_DATA, status = FETCH_STATUS.LOADING } = useFetcher(
```
why do we need a default value here?

kpatticha replied:
hmm, it was there before I just updated the hardcoded `status = "loading"` with the constant but I believe we don't need it

dgieselaar commented on x-pack/plugins/apm/public/hooks/use_environments_fetcher.tsx:
```diff
@@ -17,10 +17,16 @@ function getEnvironmentOptions(environments: string[]) {
-  return [ENVIRONMENT_ALL, ...environmentOptions];
+  return [
+    ENVIRONMENT_ALL,
+    ...(environments.includes(ENVIRONMENT_NOT_DEFINED.value)
+      ? [ENVIRONMENT_NOT_DEFINED]
+      : []),
```
does this mean we lose the sort? is the list sorted usually, and is it important? (i'd say it's okay to not care much for sorting here, as long as we've considered the implications).

kpatticha replied:
The environment list will be displayed as it used to be, the sorting remains unchanged. 

We didn't specify a sort option for the [endpoint](https://github.com/elastic/kibana/blob/main/x-pack/plugins/apm/server/routes/environments/get_environments.ts#L23) so I think we can discuss the sorting in another ticket

dgieselaar commented on x-pack/plugins/apm/public/context/environments_context/environments_context.tsx:
```diff
@@ -39,7 +45,7 @@ export function EnvironmentsContextProvider({
   const { start, end } = useTimeRange({ rangeFrom, rangeTo, optional: true });
 
-  const { environments, status } = useEnvironmentsFetcher({
+  const { environments, status, environmentOptions } = useEnvironmentsFetcher({
```
I realize this was in there before. But I'm not sure if `useEnvironmentsFetcher` should return `environmentOptions`. That seems like a responsibility of the component that displays the suggestions. Not a blocker though.

kpatticha committed:
Fix types

dgieselaar commented on x-pack/plugins/apm/public/components/shared/environment_select/index.tsx:
```diff
@@ -0,0 +1,103 @@
+      options={options}
+      selectedOptions={selectedOptions}
+      onChange={(changedOptions) => handleChange(changedOptions)}
+      onSearchChange={debounce(setSearchValue, 500)}
```
I'm not sure if this works the way you expect it to, maybe by accident, but this creates a new debounced handler every render. You probably want to wrap it in `useMemo()`.

kpatticha replied:
fixed: https://github.com/elastic/kibana/pull/126679/commits/13d49f5f799a10b15b5dd5474f521a91e7be2893

dgieselaar commented on x-pack/plugins/apm/public/components/shared/environment_select/index.tsx:
```diff
@@ -0,0 +1,103 @@
+      singleSelection={{ asPlainText: true }}
+      options={options}
+      selectedOptions={selectedOptions}
+      onChange={(changedOptions) => handleChange(changedOptions)}
```
what's the difference between onChange and onSearchChange? what's the reason one is debounced while the other is not?

kpatticha replied:
The `onSearchChange` fires when you start typing on the search box and this one is debounced cause I don't want to send a request to the backend on every single type. 

On the other side `onChange` fires only when you select an option from the dropdown and since the component is singleSelection (allows the user to select a single option) I don't think it needs debounce

dgieselaar replied:
Thanks! Can you rename the handler to `onSelect` or `onCommit` to make this difference more obvious?

dgieselaar commented on x-pack/plugins/apm/public/components/shared/environment_select/index.tsx:
```diff
@@ -0,0 +1,103 @@
+  environmentOptions: Array<EuiComboBoxOptionOption<string>>;
+  status: FETCH_STATUS;
+  serviceName?: string;
+  start?: string;
+  end?: string;
```
can we make these required?

kpatticha replied:
we can make the ranges required if we set the default values for ranged `start` and `end` in the consuming component. 

Now I have them optional because I get them from the params and handle the a [default values](https://github.com/elastic/kibana/pull/126679/files#diff-d75a3e48748c8c5022fdcb98f1cd9766a9ca02d7af80531f174c09a5480a1b78R62) here but it seems a little bit confusing. 

Which way do you think it will be cleaner. 

- the consuming app sets the default values (24h-now) and passes the ranged to `EnvironmentSelect` which are required
- the `EnvironmentSelect` component should use `useApmParams` to get the ranges and handles the default values accordingly

dgieselaar replied:
the first option sounds good to me. The consuming component should decide what the time range is because it knows in what context `EnvironmentSelect` is being used.

dgieselaar commented on x-pack/plugins/apm/public/components/shared/environment_select/index.tsx:
```diff
@@ -0,0 +1,103 @@
+    label: getEnvironmentLabel(environment),
+  };
+
+  const [selectedOptions, setSelectedOptions] = useState(
+    defaultOption ? [defaultOption] : []
+  );
+
+  const handleChange = useCallback(
+    (changedOptions: Array<EuiComboBoxOptionOption<string>>) => {
+      setSelectedOptions(changedOptions);
+
+      if (changedOptions.length === 1 && changedOptions[0].value) {
+        onChange(changedOptions[0].value);
+      }
+    },
+    [onChange]
+  );
```
I'm not sure if we need state here? I would expect this component to take an `environment` prop and an `onChange` prop, and have the consuming component handle state changes / synchronisation with the URL.

In fact, I think there's a bug: if you select an environment, and then use browser navigation to go back, it doesn't update the selected options.

kpatticha replied:
good catch 😮  I just tried it and there is a bug as you described 🐛

kpatticha replied:
fixed here: https://github.com/elastic/kibana/pull/126679/commits/31e00364e3f99960b50a3a157b1c8d2ebedb83ac

kpatticha committed:
State should be handled by consuming component

kpatticha committed:
Use searchAggregatedTransactions

kpatticha committed:
Make ranges required for /suggestion endpoint

- the client set a default 24h-now if no URL is set

kpatticha committed:
Clean up files by separating the helper

kpatticha committed:
Fix debounce handler

- remove unused initial state

kpatticha committed:
Merge remote-tracking branch 'upstream/main' into conv-env-filter

dgieselaar commented on x-pack/plugins/apm/common/anomaly_detection/get_preferred_service_anomaly_timeseries.ts:
```diff
@@ -18,7 +18,7 @@ export function getPreferredServiceAnomalyTimeseries({
 }: {
   environment: Environment;
-  environments: Environment[];
+  environments: string[];
```
this might be a code smell.. why is it needed? (it's useful because it signals that an environment has been validated, to not be an empty string)

dgieselaar commented on x-pack/plugins/apm/public/components/shared/environment_filter/index.tsx:
```diff
@@ -34,77 +27,23 @@ function updateEnvironmentUrl(
-  status: FETCH_STATUS;
-}) {
+  const { environment, serviceName, environmentOptions, status, start, end } =
+    useEnvironmentsContext();
```
I think I mentioned this before, but I can no longer find it - I don't think `useEnvironmentsContext` should return all this information. I don't see a compelling reason for `serviceName`, `environmentOptions` and `start` and `end` to be there. It's better for `useEnvironmentsContext` to only return what is necessary, to separate concerns.

dgieselaar commented on x-pack/plugins/apm/public/components/shared/environment_select/index.tsx:
```diff
@@ -0,0 +1,103 @@
+  onChange,
+}: {
+  environment: Environment;
+  environmentOptions: Array<EuiComboBoxOptionOption<string>>;
```
Maybe this should be named `availableEnvironments: Environment[]`? the option formatting sounds like it's an implementation detail of `EnvironmentSelect`.

dgieselaar commented on x-pack/plugins/apm/public/components/shared/suggestions_select/index.tsx:
```diff
@@ -48,11 +55,16 @@ export function SuggestionsSelect({
+          query: {
+            fieldName,
+            fieldValue: searchValue,
+            start: start ?? moment().subtract(24, 'h').toISOString(),
```
Preferably we don't have defaults here either, but always require it and let the consuming component set it.

kpatticha committed:
Merge branch 'main' of github.com:elastic/kibana into conv-env-filter

kpatticha committed:
Add option to run APM cypress tests in the flaky test runner

spalger reviewed (approved):
Flaky test runner changes LGTM, though it would be great if you used the `editorconfig` plugin for your editor so that it didn't unnecessarily drop the final newline in this file.

kpatticha committed:
Merge branch 'main' of github.com:elastic/kibana into conv-env-filter

kpatticha committed:
Clean up environment context

- context will be responsible only for returning environment(s) info
- removing additional info not related to the context

kpatticha committed:
Rename event handlers

kpatticha committed:
Consuming component sets ranges for the suggestions select

kpatticha committed:
Immediately resolve the promise with empty response in case searchValue is empty

kpatticha committed:
Fix tests

dgieselaar commented on x-pack/test/apm_api_integration/tests/suggestions/suggestions.spec.ts:
```diff
@@ -9,12 +9,14 @@ import {
   SERVICE_NAME,
   TRANSACTION_TYPE,
 } from '../../../../plugins/apm/common/elasticsearch_fieldnames';
+import archives_metadata from '../../common/fixtures/es_archiver/archives_metadata';
```
it's out of scope for this issue, but I think we should aim to use synthtrace everywhere rather than load the archives. that allows us to move away from snapshot testing to logical assertions.

dgieselaar commented on x-pack/plugins/apm/public/components/alerting/fields.tsx:
```diff
@@ -21,6 +21,9 @@ import {
 import { SuggestionsSelect } from '../shared/suggestions_select';
 import { PopoverExpression } from './service_alert_trigger/popover_expression';
 
+const start = moment().subtract(24, 'h').toISOString();
```
this needs to be inside of the render function, otherwise it's only initialized when the javascript loads in the browser.

kpatticha replied:
what I was thinking 🙈 Updating it

dgieselaar commented on x-pack/plugins/apm/public/components/app/settings/custom_link/create_edit_custom_link_flyout/filters_section.tsx:
```diff
@@ -35,6 +35,9 @@ export function FiltersSection({
   filters: Filter[];
   onChangeFilters: (filters: Filter[]) => void;
 }) {
+  const start = moment().subtract(24, 'h').toISOString();
```
same here

dgieselaar commented on x-pack/plugins/apm/public/components/shared/environment_select/index.tsx:
```diff
@@ -0,0 +1,133 @@
+  onChange: (value: string) => void;
+}) {
+  const [searchValue, setSearchValue] = useState('');
+  const serviceName = useServiceName();
```
we should avoid `useServiceName()` here. it's intended for places where `serviceName` is optional. In this case, we're guaranteed to have `serviceName` available as you require the route to be active on line 55.

dgieselaar commented on x-pack/plugins/apm/public/components/shared/environment_select/index.tsx:
```diff
@@ -0,0 +1,133 @@
+  );
+  const terms = data?.terms ?? [];
+
+  const environmentOptions = useMemo(
```
is it important that these are wrapped in `useMemo()`? We should avoid premature optimizations if possible. It might even be worse for performance if the `availableEnvironments` prop is not memoized.

kpatticha replied:
imo it's not important, I checked the profiler and we don't gain any performance benefits . I'll remove it

dgieselaar commented on x-pack/plugins/apm/public/components/shared/environment_select/index.tsx:
```diff
@@ -0,0 +1,133 @@
+    },
+  ];
+
+  const onSelect = useCallback(
```
same question here, is it important that this is memoized? AFAICT the `onChange` prop is not memoized, but recreated every render, so it will invalidate the dependencies every render anyway.

kpatticha committed:
Merge branch 'main' of github.com:elastic/kibana into conv-env-filter

kpatticha committed:
Get serviceName from useApmParams

kpatticha committed:
Fix type

kpatticha committed:
Initialize time ranges inside the components

kpatticha committed:
Remove redundant optimization

kpatticha committed:
Fix eslint import errors

kpatticha committed:
Fix cypress tests

kpatticha commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into conv-env-filter

kpatticha commented:
@dgieselaar do you mind having a look at the PR, hopefully it's getting closer to get merged 🙏

kpatticha committed:
Fix translations

kpatticha commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into conv-env-filter

dgieselaar commented:
@kpatticha there's a comment on `timeout` that I'd like to see addressed: https://github.com/elastic/kibana/pull/126679/files#r852876333

kpatticha commented:
> @kpatticha there's a comment on `timeout` that I'd like to see addressed: https://github.com/elastic/kibana/pull/126679/files#r852876333

@dgieselaar I have totally missed this comment. Thanks for pointing it out 🙏  I'll have a look now

kpatticha committed:
Merge remote-tracking branch 'upstream/main' into conv-env-filter

kpatticha committed:
Add 1500ms timeout

kpatticha committed:
Merge branch 'conv-env-filter' of github.com:kpatticha/kibana into conv-env-filter

dgieselaar reviewed (approved):
Thanks Kate, great work!

kpatticha commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into conv-env-filter
